Docs: information on schemaless sources & destinations (#33595)
airbytehq/airbyte: docs/understanding-airbyte/schemaless-sources-and-destinations.md

# "Schemaless" Sources and Destinations

> In order to run a sync, Airbyte requires a catalog, which includes a data schema describing the shape of data being emitted by the source.

Reviewer: ```suggestion
In order to run a sync, Airbyte requires a [catalog](/understanding-airbyte/airbyte-protocol#catalog), which includes a data schema describing the shape of data being emitted by the source.
```

This schema will be used to prepare the destination to populate the data during the sync.

> This document describes the options available for the subset of sources that do not have a strict schema, aka "schemaless sources".

Reviewer: ```suggestion
While having a [strongly-typed](/understanding-airbyte/supported-data-types) catalog/schema is possible for most sources, some won't have a reasonably static schema.  This document describes the options available for the subset of sources that do not have a strict schema, aka "schemaless sources".
```

## What is a Schemaless Source?

Schemaless sources are sources for which there is no requirement or expectation that records will conform to a particular pattern.

> For example, in a MongoDB database, there's no requirement that the fields in one record are the same as the fields in the next, or that the type of value in one field is the same as the type for that field in a separate record.

Reviewer 2: MongoDB has a "document" rather than a record.

Author: 👍 updated.

Similarly, for a file-based source such as S3, the files that are present in your source may not all have the same schema.

Although the sources themselves may not conform to an obvious schema, Airbyte still needs to know the shape of the data in order to prepare the destination for the records.

> For these sources, during the `discover` call, Airbyte offers two options to create the schema:

Reviewer: ```suggestion
For these sources, during the [`discover`](/understanding-airbyte/airbyte-protocol#discover) method, Airbyte offers two options to create the schema:
```

1. Dynamic schema inference.
2. A hardcoded "schemaless" schema.

### Dynamic schema inference

> If this option is selected, Airbyte will infer the schema dynamically based on the contents of the source.

Reviewer: ```suggestion
If this option is selected, Airbyte will infer the schema dynamically based on the contents of the source.  If your source's content is homogenous, we recommend this option, as the data in your destination will be typed and you can make use of schema evolution features, column selection, and similar Airbyte features which operate against the source's schema. 
```

> For MongoDB, if the dataset exceeds 10,000 documents, the schema is inferred from a random sample of up to 10,000 documents; otherwise, it is inferred from the entire dataset.

Reviewer 2: Sample size is configurable in source-mongodb: 1,000 - 100,000

Author: 👍 updated.

So if we observe that a field has an integer type in one record and a string in another, the schema will identify the field as a string.

There are a few drawbacks to be aware of:
- If your dataset is very large, the `discover` process can be very time-consuming.
- Because we may not use 100% of the available data to create the schema, your schema may not contain every field present in your records.
  Airbyte only syncs fields that are in the schema, so you may end up with incomplete data in the destination.

If your data set is very large or you anticipate that it will change often, we recommend using the "schemaless" schema.

### Schemaless schema
If this option is selected, the schema will always be `{"data": object}`, regardless of the contents of the data. During the sync, we "wrap" each record behind a key named `data`. This means that the destination receives the data with one top-level field only, and the value of the field is the entire record.
This option avoids a time-consuming or inaccurate `discover` phase and guarantees that everything ends up in your destination, at the expense of Airbyte being able to structure the data into different columns.

> ## Coming Soon

Reviewer: ```suggestion
## Future Enhancements
```

Until we prioritize this work, I don't think "soon" is the right word.

> ## Future Enhancements

Reviewer 2: Suggestion: a way to exclude fields from syncing in schemaless also, as column selection is not applicable.
This is on the radar for mongodb.

Author: 👍 added.
